### ISSUE TITLE
equation references do not work forward [was: AMS numbering causes numbered equations to be smaller, and `eqref` fails to render.]

### ISSUE THREAD

**Issue by fasiha** [pkra/mathjax-node-page#47]
I was surprised when AMS numbering & equation references, which render fine with regular browser-MathJax, have a these problems with `mjpage`.

Here’s my minimal HTML, cribbed from [MathJax docs](http://docs.mathjax.org/en/latest/tex.html):
```html
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">
</script>

For example

\begin{equation}
   e^{\pi i} + 1 = 0
\end{equation}
will be numbered, while

\begin{equation*}
   e^{\pi i} + 1 = 0
\end{equation*}
won’t be numbered (when autoNumber is "AMS").

You can use \label to give an equation an identifier that you can use to refer to it later, and then use \ref or \eqref within your document to insert the actual equation number at that location, as a reference. For example,

In equation \eqref{eq:sample}, we find the value of an interesting integral:

\begin{equation}
  \int_0^\infty \frac{x^3}{e^x-1}\,dx = \frac{\pi^4}{15}
  \label{eq:sample}
\end{equation}
includes a labeled equation and a reference to that equation.
```
When I open this in the browser, I get exactly what I expect:

![screen shot 2017-05-28 at 6 18 34 pm](https://cloud.githubusercontent.com/assets/37649/26532586/4384ae3a-43d2-11e7-868c-0ee8ffe6670e.png)

But when I put it through `mjpage` 1.4.0:
```
$ mjpage --eqno "AMS" < test.html > test2.html
```
the result looks like this:

![screen shot 2017-05-28 at 6 18 36 pm](https://cloud.githubusercontent.com/assets/37649/26532590/5f9b9ac0-43d2-11e7-93c7-61022f6464a8.png)

- The numbered equation is *smaller* than the unnumbered one, and 
- the equation reference `\eqref` is ignored, and a blank space appears where it should be.

**Comment by pkra** [comment]
Thanks for the report. Looks like an upstream issue so I've filed https://github.com/mathjax/MathJax-node/issues/333.

On an unrelated note, I wouldn't advise using labels to the right in the SVG output as SVG cannot make the label's position stretch to the right (but instead the SVG is fixed to whatever size you configure MathJax to). 

With SVG, I generally tell clients to put labels on the left and left-align block equations (possibly with indentation).

**Comment by fasiha** [comment]
Thank you! I really like this idea, so something like, `<div id="eqno-123"><span class="eqno">(123)</span> <svg………>`. Is there a way to achieve this with MathJax or would I write some custom HTML processing?

Also, any thoughts on `\eqref` in mjpage?

**Comment by pkra** [comment]
> Is there a way to achieve this with MathJax or would I write some custom HTML processing?

First off, in case it wasn't clear, you can use the default label style ("floating" to the right) with the CommonHTML output.

Second, you can set `displayAlign` and `displayIndent` from the [core](http://docs.mathjax.org/en/latest/options/hub.html) MathJax options to change alignment. For [TeX](http://docs.mathjax.org/en/latest/options/TeX.html) input, you can set `TagSide` and `TagIndent` options for the TeX input to position equation labels.

> any thoughts on \eqref in mjpage?

`\ref` works in MathJax. `\eqref` is not built into in MathJax so you'd have to write a macro yourself.

**Comment by dpvc** [comment]
FYI, `\eqref` *is* defined in MathJax (and the code to handle that has been copied here as well; see [line 101](https://github.com/pkra/mathjax-node-page/blob/master/lib/tex.js#L101) of the `tex.js` in this archive).  But it looks like mjpage will only handle backward references (not the forward references handled by MathJax).

**Comment by pkra** [comment]
> \eqref is defined in MathJax (and the code to handle that has been copied here as well; see line 101 of the tex.js in this archive).

:bowing_man: My bad -- thanks, Davide! 

> But it looks like mjpage will only handle backward references (not the forward references handled by MathJax).

Will look into it.

**Comment by pkra** [comment]
FYI the upstream issue was marked expected behavior with several suggestions available. 

So what's left is the lack of support here for references that look ahead.

**Comment by pkra** [comment]
This actually has two problems. On the one hand, I broke the TeX preprocessor's ref/eqref handling. On the other hand, we'll need some logic to look forward.

I'll make a simple fix but it won't handle all cases -- it'll just move equations with ref/eqref to the end of the processing queue. A refinement could sort the queue better (check if an expression references another) but we probably can't resolve circular references (without a lot of work).
